### ISSUE TITLE
CompactOnLaunch

### ISSUE THREAD

**Issue by diegomontoya** [realm/realm-java#3739]
#### Goal

End users need to know the amount of dirty/recoverable space within the realm file to 

- Improve scheduling of compaction() call on a realm 
- Debug run away concurrent read/writes from different threads causing explosion of the realm file due to the duplicate of space used in multi-versioning. 

#### Expected Results

ReclaimableSpace() API which tells us in bytes, or approximate bytes if 100% accuracy cannot be achieved, of recoverable space in the realm file that can be trimmed in compaction.

**Comment by cmelchior** [comment]
Hmm, interesting suggestion. I'm not sure it would be that useful for debugging run away concurrent read/writes since the relationship between file size and available space is not 1:1

In 2.0 we added `Realm.getGlobalInstanceCount(config)` and `Realm.getLocalInstanceCount(config)` which should help with debugging run-away instances.

I created https://github.com/realm/realm-core/issues/2281 for Realm Core to expose the available space information.

**Comment by cmelchior** [comment]
Core merged an API that might help with this here: https://github.com/realm/realm-core/pull/2338

**Comment by Zhuinden** [comment]
Oh then you already know about https://github.com/realm/realm-core/pull/2338 just saw it a second ago

**Comment by cmelchior** [comment]
Cocoa have merged support for this here: https://github.com/realm/realm-cocoa/pull/4755/files

The equivalent Java API should look something like this:

```
// Cocoa example
RealmConfiguration config = new RealmConfiguration.Builder()
  .compactOnLaunch(new CompactOnLaunchCallback() {
     @Override
     public boolean shouldCompact(long totalBytes, long usedBytes) {
       // Compact if the file is over 500KB in size and less than 20% 'used'
       long minSize = 500 * 1024;
       return (totalBytes > minSize) && (usedBytes / (double) totalBytes) < 0.2;     } 
  })
  .build();

// Possible helper method for the simple cases
RealmConfiguration config = new RealmConfiguration.Builder()
  .compactOnLaunch(0.2) // Compact if it would reclaim at least 20%, no matter the size
  .build(); 
``

`compactOnLaunch` do not work for synced Realms.

**Comment by diegomontoya** [comment]
What about a more generic version.

```java
RealmConfiguration config = new RealmConfiguration.Builder()
  .autoCompact(new AutoCompactListener() {
     @Override
     public boolean shouldCompact(boolean isLaunch, long totalBytes, long usedBytes) {
       // Compact if the file is over 500KB in size and less than 20% 'used'
       long minSize = 500 * 1024;
       return (totalBytes > minSize) && (usedBytes / (double) totalBytes) < 0.2;     } 
  })
  .build();
```
This would fit the compact on launch needs and any future run-time compaction feature expansion.

**Comment by Zhuinden** [comment]
This listener looks similar to the sample provided for the cocoa version of this feature https://realm.io/news/realm-objc-swift-2.6/

**Comment by cmelchior** [comment]
With #4570 merged, this issue doesn't have any more blockers

**Comment by cmelchior** [comment]
@diegomontoya By `isLaunch` you mean this boolean is true the first time the Realm is opened right (GlobalCount == 0) ? I could see that being useful, yes.

**Comment by cmelchior** [comment]
Merged in #4857
